<a href="https://colab.research.google.com/github/swarnava-96/Recommendation-Systems/blob/main/Recommender_Engine_in_Retail_Scenario.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Market Basket Ananlysis**

In [2]:
# Loading the necessary libraries
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

In [3]:
# Reading Data From Web
myretaildata = pd.read_excel('http://archive.ics.uci.edu/ml/machine-learning-databases/00352/Online%20Retail.xlsx')
myretaildata.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

### Feature Engineering

In [4]:
# Lets see the shape of the dataset
myretaildata.shape

(541909, 8)

In [5]:
# Lets see some informations of the data
myretaildata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [7]:
# Removing spaces from the begining and end
myretaildata["Description"] = myretaildata["Description"].str.strip()
# Lets remove null values in invoice
myretaildata.dropna(axis = 0, subset = ["InvoiceNo"], inplace = True)
# Converting "InvoiceNo" to string as it is a requirement of the model
myretaildata["InvoiceNo"] = myretaildata["InvoiceNo"].astype("str")
# Removing credit transactions
myretaildata = myretaildata[~myretaildata["InvoiceNo"].str.contains("C")]
# Lets see the data now
myretaildata.head()

InvoiceNo StockCode  ... CustomerID         Country
0    536365    85123A  ...    17850.0  United Kingdom
1    536365     71053  ...    17850.0  United Kingdom
2    536365    84406B  ...    17850.0  United Kingdom
3    536365    84029G  ...    17850.0  United Kingdom
4    536365    84029E  ...    17850.0  United Kingdom

[5 rows x 8 columns]

In [8]:
# Lets check the shape now
myretaildata.shape

(532621, 8)

In [9]:
# Lets see the number of unique countries data
myretaildata["Country"].value_counts()

United Kingdom          487622
Germany                   9042
France                    8408
EIRE                      7894
Spain                     2485
Netherlands               2363
Belgium                   2031
Switzerland               1967
Portugal                  1501
Australia                 1185
Norway                    1072
Italy                      758
Channel Islands            748
Finland                    685
Cyprus                     614
Sweden                     451
Unspecified                446
Austria                    398
Denmark                    380
Poland                     330
Japan                      321
Israel                     295
Hong Kong                  284
Singapore                  222
Iceland                    182
USA                        179
Canada                     151
Greece                     145
Malta                      112
United Arab Emirates        68
European Community          60
RSA                         58
Lebanon 

In [12]:
# Seperating the transactions for Germany only
mybasket = (myretaildata[myretaildata['Country'] =="Germany"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

In [13]:
# Lets see our transaction basket
mybasket.head()

Description  10 COLOUR SPACEBOY PEN  ...  ZINC WILLIE WINKIE  CANDLE STICK
InvoiceNo                            ...                                  
536527                          0.0  ...                               0.0
536840                          0.0  ...                               0.0
536861                          0.0  ...                               0.0
536967                          0.0  ...                               0.0
536983                          0.0  ...                               0.0

[5 rows x 1695 columns]

In [14]:
# Converting all positive values to 1 and everyhing else to zero
def my_encode_units(x):
  if x <= 0:
    return 0
  if x >= 1:
    return 1

my_basket_sets = mybasket.applymap(my_encode_units)
my_basket_sets.drop("POSTAGE", axis = 1, inplace = True)

### Model Training

In [16]:
# Generating frequent itemsets
my_frequent_itemsets = apriori(my_basket_sets, min_support = 0.07, use_colnames = True) 

In [17]:
# Generating rules
my_rules = association_rules(my_frequent_itemsets, metric = "lift", min_threshold = 1)

In [18]:
# Lets see the top 100 rules
my_rules.head(100)

antecedents  ... conviction
0    (PLASTERS IN TIN WOODLAND ANIMALS)  ...   1.640006
1  (ROUND SNACK BOXES SET OF4 WOODLAND)  ...   1.237951
2   (ROUND SNACK BOXES SET OF 4 FRUITS)  ...   4.529540
3  (ROUND SNACK BOXES SET OF4 WOODLAND)  ...   1.814509
4                  (SPACEBOY LUNCH BOX)  ...   2.365427
5  (ROUND SNACK BOXES SET OF4 WOODLAND)  ...   1.256018

[6 rows x 9 columns]

### Recommendations

In [19]:
my_basket_sets["ROUND SNACK BOXES SET OF4 WOODLAND"].sum()

112

In [20]:
my_basket_sets['SPACEBOY LUNCH BOX'].sum()

47

We can say out of 112 times "ROUND SNACK BOXES SET OF4 WOODLAND" purchased, 47 times "SPACEBOY LUNCH BOX" was also purchased along with it. So, when a customer purchase a "ROUND SNACK BOXES SET OF4 WOODLAND" in future, we can recommend him/her to purchase "SPACEBOY LUNCH BOX" along with it.

In [21]:
#Filtering rules based on condition
my_rules[(my_rules['lift'] >= 3) &
       (my_rules['confidence'] >= 0.3)]

antecedents  ... conviction
2   (ROUND SNACK BOXES SET OF 4 FRUITS)  ...   4.529540
3  (ROUND SNACK BOXES SET OF4 WOODLAND)  ...   1.814509

[2 rows x 9 columns]